### PR TITLE
Allow specifying custom code generation settings

### PR DESCRIPTION
**Pull request by jhrcek**
I'm playing with ways that would allow resolving https://github.com/Holmusk/elm-street/issues/129
while keeping the old behavior (stripping type name prefix) by default.

The important thing is that generated Elm encoders and decoders have to be compatible with ToJSON / FromJSON instances of Haskell types used across Haskell + Elm. This means that both Elm instance and ToJSON / FromJSON must use the same (what I call) `CodeGenSettings` that carry the "field label modifier" function (which for now is the only aspect of code generation that we allow to override).

See Type.hs for an example of how this could be used.

I'll need to think if this could be done more nicely over the next few days, but feel free to provide feedback if any.

### CODE REVIEW

**Review by turboMaCk — APPROVED** [review]
Looks great :+1:

probably could use some documentation updates though

**Comment by jhrcek** [comment]
Thanks, docs updated.

File: test/golden/oneType.json [hunk 1]
```diff
@@ -100,4 +100,4 @@
             "contents": []
         }
     ]
-}
+}
```
Comment (turboMaCk) [inline test/golden/oneType.json:103]: no new line? :eyes:

**Comment by turboMaCk** [comment]
Is anything blocking merge of this?

**Comment by jhrcek** [comment]
We didn't get any feedback from @silky who actually requested that change.
Let me revisit this PR and make it mergeable within the next few days if we don't hear from him.

**Comment by silky** [comment]
Apologies for never getting around to responding to this; in the end I changed jobs since I needed it :)

But at a glance, it definitely does what I wanted, and I'll probably have cause to use it again at some point in the future; so that's great! Thanks :) 🙏🏻

**Review by nitinprakash96 — APPROVED** [review]

File: src/Elm/Aeson.hs [hunk 2]
```diff
@@ -73,13 +71,13 @@ it is decoded it the following way for each of the specified types:
...
 >>> instance FromJSON User where parseJSON = elmStreetParseJson
->>> decode @User "{ \"name\": \"John\", \"age\": 42 }"
+>>> decode @User "{\"age\":42,\"name\":\"John\",\"tag\":\"User\"}"
 Just (User {userName = "John", userAge = 42})
 
 
 >>> data VeryLongType = VeryLongType { vltName :: String, vltAge :: Int } deriving (Generic, Show)
 >>> instance FromJSON VeryLongType where parseJSON = elmStreetParseJson
->>> decode @VeryLongType "{ \"name\": \"John\", \"age\": 42 }"
+>>> decode @VeryLongType "{\"age\":42,\"name\":\"John\",\"tag\":\"VeryLongType\"}"
```
Comment (turboMaCk) [inline src/Elm/Aeson.hs:80]: doesn't seem that long to me :eyes: `ProjectContractChargingPeriodProjectAccountReferenceVMFactoryBuilderStrategy`

**Pull request merged**